lib/session.g

```diff
@@ -25,10 +25,15 @@ InstallAtExit( function()
 end);
 
 BIND_GLOBAL("PROGRAM_CLEAN_UP", function()
-    local f;
+    local f, funcs;
     if IsBound( GAPInfo.AtExitFuncs ) and IsList( GAPInfo.AtExitFuncs ) then
-        while not IsEmpty(GAPInfo.AtExitFuncs) do
-            f := Remove(GAPInfo.AtExitFuncs);
+        if IsHPCGAP then
+            funcs := FromAtomicList(GAPInfo.AtExitFuncs);
```

Comment:
(I think I changed this last, but I am looking just on github). The original code was designed so in the event new things got pushed onto the atexitfuncs queue during program_clean_up, they would still get run. With this new design that would be lost, at least in hpcgap. I have no idea if this is a real problem, but thought I'd mention it.

Reply:
Good point. I think no sane AtExit function should modify `GAPInfo.AtExitFuncs` but of course some people are evil. Anyway, I'd be fine with explicitly not supporting it. Best would of course be to explicitly prevent it -- say by inserting a call to `MakeImmutable(GAPInfo.AtExitFuncs)` before iterating over it?

@rbehrends any other thoughts?

Reply:
I think the cleanest solution would actually be to modify `InstallAtExit()` to properly coordinate with the `PROGRAM_CLEAN_UP()` logic. Right now, there is no way to cleanly implement `Remove()` for atomic lists; we could instead surround it with locking logic so that we can't simultaneously add and remove elements from the list.

This would possibly break code that manipulates `GAPInfo.AtExitFuncs` directly, however.